File: lab3/python/lab3.py
#!/usr/bin/python2.7

import math
import matplotlib.pyplot as plt
a = 6378137.0
f = 1.0 / 298.257222101
b = a - a * f
e = math.sqrt((a**2 - b**2) / a**2)

Lat = []
M_80 = []
N_80 = []
Ravg_80 = []
RG_80 = []

for i in range(-90, 91):
    Lat.append(i)
    M = a * (1 - e**2) / math.sqrt((1 - e**2 * (math.sin(math.radians(i)))**2)**3)
    M_80.append(M)
    N = a / math.sqrt(1 - e**2 * (math.sin(math.radians(i)))**2)
    N_80.append(N)
    Ravg_80.append(1.0 * (M + N) / 2)
    RG_80.append(math.sqrt(M * N))
    
a = 6378160.0;
f = 1.0 / 298.247167427;
b = a - a * f
e = math.sqrt((a**2 - b**2) / a**2)
M_67 = []
N_67 = []
Ravg_67 = []
RG_67 = []
dM = []
dN = []
dRavg = []
dRG = []


for i in range(-90, 91):
    M = a * (1 - e**2) / math.sqrt((1 - e**2 * (math.sin(math.radians(i)))**2)**3)
    M_67.append(M)
    N = a / math.sqrt(1 - e**2 * (math.sin(math.radians(i)))**2)
    N_67.append(N)
    Ravg_67.append(1.0 * (M + N) / 2)
    RG_67.append(math.sqrt(M * N))
    
for i in range(181):
    dM.append(M_67[i] - M_80[i])
    dN.append(N_67[i] - N_80[i])
    dRavg.append(Ravg_67[i] - Ravg_80[i])
    dRG.append(RG_67[i] - RG_80[i])
    
fig, ax = plt.subplots()

ax.plot(Lat, dM, label = "Radii of curvative on meridian plane")
ax.plot(Lat, dN, label = "Radii of curvative on prime vertical plane")
ax.plot(Lat, dRavg, label = "Average of curvative")
ax.plot(Lat, dRG, label = "Gaussian curvative")

#ax.plot(Lat, M_80, label = "Radii of curvative on meridian plane")
#ax.plot(Lat, N_80, label = "Radii of curvative on prime vertical plane")
#ax.plot(Lat, Ravg_80, label = "Average of curvative")
#ax.plot(Lat, RG_80, label = "Gaussian curvative")

ax.annotate(r'Radii of curvative on meridian plane', xy=(Lat[111], dM[111]), 
            xytext=(30, -80), textcoords='offset points', va='center',
            arrowprops=dict(arrowstyle='->'))
#
ax.annotate(r'Radii of curvative on prime vertical plane', xy=(Lat[130], dN[130]), 
            xytext=(-140, 80), textcoords='offset points', va='center',
            arrowprops=dict(arrowstyle='->'))
            
ax.annotate('Average of curvative\n               &\n Gaussian curvative', xy=(Lat[120], dRavg[120]), 
            xytext=(-140, 50), textcoords='offset points', va='center',
            arrowprops=dict(arrowstyle='->'))
            
#ax.annotate('Average of curvative, Ravg_80[0]=%.6f' % Ravg_80[0], xy=(-90, Ravg_80[0]), 
#            xytext=(-70, 80), textcoords='offset points', va='center',
#            arrowprops=dict(arrowstyle='->'))
#            
#ax.annotate('Gaussian curvative, RG_80[0]=%.6f' % RG_80[0], xy=(-90, RG_80[0]), 
#            xytext=(-70, -80), textcoords='offset points', va='center',
#            arrowprops=dict(arrowstyle='->'))


plt.xlabel("Latitude  (Unit: degree)")
plt.ylabel("Radii of curvature  (Unit: m)")
plt.show()

File: lab3/matlab/lab3.m
%GRS67 ellipsodial parameter values
a_67 = 6378160.0;
f_67 = 1.0 / 298.247167427;
b_67 = a_67 - a_67 *f_67;
e_67 = sqrt((a_67^2 - b_67^2) / (a_67^2));

%GRS80 ellipsodial parameter values
a_80 = 6378137.0;
f_80 = 1.0 / 298.257222101;
b_80 = a_80 - a_80 *f_80;
e_80 = sqrt((a_80^2 - b_80^2) / (a_80^2));

Latitude = zeros(181, 1);
M_67 = zeros(181, 1); N_67 = zeros(181, 1); Ravg_67 = zeros(181, 1); RG_67 = zeros(181, 1);
M_80 = zeros(181, 1); N_80 = zeros(181, 1); Ravg_80 = zeros(181, 1); RG_80 = zeros(181, 1);
index = 1;
for i = -90:1:90
    Latitude(index, 1) = i;
    M_67(index, 1) = (a_67 * (1 - e_67^2)) / (1 - e_67^2 * (sin(i * (pi / 180)))^2)^(3.0 / 2.0);
    N_67(index, 1) = a_67 / sqrt(1 - e_67^2 * (sin(i * (pi / 180)))^2);
    Ravg_67(index, 1) = (M_67(index, 1) + N_67(index, 1)) / 2.0;
    RG_67(index, 1) = sqrt(M_67(index, 1) * N_67(index, 1));
    M_80(index, 1) = (a_80 * (1 - e_80^2)) / (1 - e_80^2 * (sin(i * (pi / 180)))^2)^(3.0 / 2.0);
    N_80(index, 1) = a_80 / sqrt(1 - e_80^2 * (sin(i * (pi / 180)))^2);
    Ravg_80(index, 1) = (M_80(index, 1) + N_80(index, 1)) / 2.0;
    RG_80(index, 1) = sqrt(M_80(index, 1) * N_80(index, 1));
    index = index + 1;
end

format long g
A = [Latitude, M_67, N_67, Ravg_67, RG_67, M_80, N_80, Ravg_80, RG_80];
disp(A);

%hold on;
%plot(Latitude, M_80);
%plot(Latitude, N_80);
%plot(Latitude, Ravg_80);
%plot(Latitude, RG_80);
%hold off;
